### PR TITLE
Allow for integration of nextcloud

### LINKED ISSUE #1854: OAuth: NextCloud Login mit Hitobito ermöglichen
Als Hitobito-Nutzer, der auch Nextcloud nutzt
möchte ich mich bei NextCloud via hitobito einloggen
und Rolleninformationen übergeben können,
um die Berechtigung für alle meine Vereinsmitglieder konsistent verwalten zu können.

NextCloud kann mit dem [OpenID User Backend](https://github.com/nextcloud/user_oidc) die Authentication via OAuth z.B. gegen hitobito vornehmen.

Idealerweise kann man die hitobito-Rollen NextCloud so mitteilen, dass daraus NextCloud-Gruppen mit entsprechenden Berechtigungen werden. 

# Tech-Spec

- Ziel ist die zentrale Nutzer- und Berechtigungsverwaltung
- schöner Nebeneffekt ist, dass man nur noch seinen hitobito-Login braucht
- die initiale Verbindung (Providerconfig in hitobito und NextCloud, Gruppen in NextCloud) erfolgt manuell bis es so oft passiert, dass wir es dokumentieren und später automatisieren können.

# ToDo

- [x] NextCloud-Testinstanz bereitstellen, um Integration zu validieren
- [x] Möglichkeit finden, um hitobito-Rollen an NextCloud weiterzugeben. (`user_oidc` mit PR 502)
- [x] Rollen-Klassen erweitern, um Gruppenmapping zu unterstützen (ein bestimmte Rolle innerhalb einer bestimmten Ebene bekommt direkt Gruppenname zugewiesen)
- [x] Gruppen erweitern, um Link zu NextCloud aufzunehmen (String `nextcloud_url` auf `groups`) (nur Layergruppen)
- [x] OAuth-Scope hinzufügen, um die für NextCloud notwendigen Daten zurückzugeben.
- [x] Doku hinzufügen, um tatsächliche Integration festzuhalten. 
- [x] Specs schreiben
- [x] Mit angemessener Rolle "durchklicken"
- [x] CHANGELOG-Eintrag unter "unreleased" unten hinzufügen

### PR DESCRIPTION
**Pull request by kronn**
Everything is hidden behind the FeatureGate `groups.nextcloud`. This is disabled by default.

Once activated, you can

- add a nextcloud-url to any layer-group and have it displayed on every group in the layer
- use a nextcloud-scope for oauth-apps. This allows to pass user information and a list of groups to nextcloud to provision users there.
- configure roles to either map to a fixed groupname or a dynamic group, depending on the group you have the role on.

Apart from that, I added a rake-task to determine the correct configuration quicker. The idea is to have checks for everything that requires extra-steps to be configured, collected in `rake hitobito:check_config`.

Closes #1854

### CODE REVIEW

**Comment by Michael-Schaer** [comment]
Das sieht spannend aus!

Wenn es da Möglichkeiten gibt, das bei uns zu aktivieren, haben wir die Ohren offen :)

File: db/migrate/20230115160409_add_nextcloud_link_to_groups.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddNextcloudLinkToGroups < ActiveRecord::Migration[6.1]
```
Comment (TheWalkingLeek) [inline db/migrate/20230115160409_add_nextcloud_link_to_groups.rb:1]: license

**Review by TheWalkingLeek — APPROVED** [review]
Looks good to me, maybe someone with more Oauth knowledge could look over the Oauth parts just to be sure :)

File: app/models/role.rb [hunk 2]
```diff
@@ -129,6 +137,15 @@ def archived?
     archived_at.present?
   end
 
+  def nextcloud_group
+    FeatureGate.assert! 'groups.nextcloud'
+
+    case (setting = self.class.nextcloud_group)
+    when String then { 'gid' => "hitobito-#{setting}", 'displayName' => setting }
```
[the remark this answers is not in the document]
Comment (kronn) [reply app/models/role.rb:144]: As far as I understand it `gid` is the group id. I could also only pass the names and the nextcloud-plugin that does the mapping generates a sha256 from the name to have a unique ID. Since the group-names can change, I prefer this.

File: db/migrate/20230115160409_add_nextcloud_link_to_groups.rb (hunk 1 above)
Comment (kronn) [reply db/migrate/20230115160409_add_nextcloud_link_to_groups.rb:8]: oops. updated.

**Comment by kronn** [comment]
I'd like your OAuth-expertise on this one, @carlobeltrame. Could you review this?

File: app/models/role.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlobeltrame) [reply app/models/role.rb:144]: Do you intend to serve two different use cases with the boolean vs. string? If I specify a string (e.g. "Trainers" in SKV), and there are more than one group of that type in hitobito, all people from all these groups will go to the same nextcloud group. On the other hand, if I specify just `true`, all people from these different trainer groups will be put into separate nextcloud groups. Seems confusing to me as a developer: usually `true` is just a shorthand for the default behaviour in such DSLs, but here there is no way to acheive with a string the same behaviour as `true` produces.
If both ways are definitely needed, how about letting the role class define a method or Proc class attr instead, which can generate the nextcloud group id and display name based on the role instance? This would also be more extensible for the future, if we e.g. ever need to put the user into multiple nextcloud groups based on just one hitobito role.

File: app/views/groups/_attrs.html.haml [hunk 3]
```diff
@@ -13,6 +13,8 @@
       = render_attrs(entry, :self_registration_link, :self_registration_notification_email)
     = render_attrs(entry, :type_name)
     = render_present_attrs(entry, *entry.used_attributes(:description))
+    - if FeatureGate.enabled?('groups.nextcloud')
+      = render_present_attrs(entry, :nextcloud_url)
```
Comment (carlobeltrame) [inline app/views/groups/_attrs.html.haml:17]: Wouldn't our `used_attributes` mechanism also apply here? Couldn't you add the nextcloud_url to the group's used_attrs only if the feature gate allows it? Are there no downsides to nextcloud_url not being in the used_attributes?

File: config/initializers/doorkeeper_openid_connect.rb [hunk 4]
```diff
@@ -60,7 +67,7 @@
   # end
 
   claims do
-    claim(:email, scope: :email)     { |resource_owner| resource_owner.email }
+    claim(:email, scope: :email, response: [:user_info, :id_token]) { |resource_owner| resource_owner.email }
```
Comment (carlobeltrame) [inline config/initializers/doorkeeper_openid_connect.rb:70]: Why is this change necessary? Shouldn't the user_info and id_token be linked to the `openid` scope instead?

File: config/locales/models.de.yml [hunk 5]
```diff
@@ -1107,6 +1108,7 @@ de:
       people: Lesen aller Personen, auf die du Zugriff hast
       invoices: Lesen aller Rechnungen, auf die du Zugriff hast
       mailing_lists: Lesen aller Abos, auf die du Zugriff hast
+      nextcloud: Lesen der eingestellten Nextcloud-Gruppen und des Anzeigenamens
```
Comment (carlobeltrame) [inline config/locales/models.de.yml:1111]:
```suggestion
      nextcloud: Lesen deiner Nextcloud-Gruppen und deines Anzeigenamens
```
Aus User-Sicht sind die Nextcloud-Gruppen keine Einstellungsmöglichkeit.

File: config/settings/test.yml [hunk 6]
```diff
@@ -20,3 +20,7 @@ email:
   sender: "hitobito <noreply@localhost>"
   mass_recipient: "hitobito <noreply@localhost>"
   list_domain: localhost
+
+groups:
+  nextcloud:
+    enabled: true
```
Comment (carlobeltrame) [inline config/settings/test.yml:26]: Remove? Or do tests not inherit the default settings?

File: doc/development/08_oauth.md [hunk 7]
```diff
@@ -30,6 +30,7 @@ Different external applications need different types of data about their users.
 | `people`        | people API endpoints                      |
 | `invoices`      | invoice API endpoints                     |
 | `mailing_lists` | mailing list API endpoints                |
+| `nextcloud`     | maps roles to nextcloud-groups            |
```
Comment (carlobeltrame) [inline doc/development/08_oauth.md:33]:
```suggestion
| `nextcloud`     | nextcloud groups based on user's roles    |
```
A mapping would intuitively mean a hash to me, but this scope just outputs a flat list of groups.

File: lib/tasks/hitobito.rake [hunk 8]
```diff
@@ -34,4 +34,28 @@ namespace :hitobito do
...
+  namespace :checks do
+    task oauth: :environment do
+      signing_key = Settings.oidc.signing_key.join.presence
+
+      if signing_key.nil?
+        puts <<~MESSAGE
+          ❌ OAuth not correctly configured
+
+            JWT Signing Key missing.
+            This key is needed to OAuth to work.
```
Comment (carlobeltrame) [inline lib/tasks/hitobito.rake:52]:
```suggestion
            This key is needed for OAuth to work.
```

File: app/models/role.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kronn) [reply app/models/role.rb:144]: Currently, there are the two use-cases you identified. For the SKV we have several Fachkommissionen, which should map to different nextcloud-groups.

The current method could be extended to support a Symbol (pointing to a method) or a Proc (which could be called). I will try if this is as easy as I think it is.

File: config/settings/test.yml (hunk 6 above)
Comment (kronn) [reply config/settings/test.yml:26]: The value here is different from the default settings. 

The default is to have it **disabled**, but for tests I want it **enabled**.

File: config/initializers/doorkeeper_openid_connect.rb (hunk 4 above)
Comment (kronn) [reply config/initializers/doorkeeper_openid_connect.rb:70]: The default for `response` is `[:user_info]`. This omits the email from the generated id-token. In order to fully provision the user in nextcloud (and allow for mail-notifications), the email needs to passed in the id-token. 

I did not find a way to have one claim have different response-types depending on the scope. Also, the OpenID-scope does not pass the email.

What am I missing here?

File: app/models/role.rb [hunk 9]
```diff
@@ -54,6 +54,14 @@ class Role < ActiveRecord::Base
   class_attribute :merge_excluded_attributes
   self.merge_excluded_attributes = []
 
+  FeatureGate.if('groups.nextcloud') do
```
[the remark this answers is not in the document]
Comment (kronn) [reply app/models/role.rb:57]: Good point.

File: app/views/groups/_attrs.html.haml (hunk 3 above)
Comment (kronn) [reply app/views/groups/_attrs.html.haml:17]: Good point, I'll try that.

File: app/models/role.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (kronn) [reply app/models/role.rb:57]: I was not able to extract this into a concern, despite following the usual pattern. Since it does not bother me too much personally, I will not do this now.

File: app/models/role.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kronn) [reply app/models/role.rb:154]: It was as easy, though I am unsure what "can of worms" I opened with this flexibility. It's not a pandoras box, but can break the integration in a very subtle and hard-to-debug way. I tried to document the class_attribute better, in order to at least provide some help.

File: app/decorators/group_decorator.rb [hunk 10]
```diff
@@ -97,6 +97,10 @@ def subgroup_ids
                            .pluck(:id)
   end
 
+  def nextcloud_url
+    layer_group.nextcloud_url
```
[the remark this answers is not in the document]
Comment (kronn) [reply app/decorators/group_decorator.rb]: I thought about it and implemented it.

File: app/views/groups/_attrs.html.haml (hunk 3 above)
Comment (kronn) [reply app/views/groups/_attrs.html.haml]: The views are now free from this specific FeatureGate.

I think that in most cases we can and should restrict the FeatureGate to ruby-files.

Good catch.

File: config/initializers/doorkeeper_openid_connect.rb (hunk 4 above)
Comment (kronn) [reply config/initializers/doorkeeper_openid_connect.rb:70]: I could try and evaluate the scope associated with the access_token, but Doorkeeper only selects the claims which belong to the scopes associated with the access_token. I could then try an see with other scopes (and therefore claims) are present and determine further steps. 

Each claim can only be linked to one scope (at least in Doorkeeper/OIDC). Linking a central piece of the user to a not always used scope is not useful to me.

OTOH, the id_token only contains data from the claims are covered by the authorized scopes. So, if you do not want to pass your email to your OIDC-app, the scope "email" should not be part of the OAuth-Application. If the consuming app need the e-mail, it will be in the application and (hopefully) authorized by the user. With this, the consuming application receives the data directly with the JWT instead of having to ask for it via the user_info-endpoint.

So, in summary, I think the change is necessary because some OAuth-consumers only evaluate the id_token, especially nextcloud. Since no additional claim is linked to the openid-scope, there is no response to configure.

**Review by carlobeltrame — APPROVED** [review]

**Review by carlobeltrame — COMMENTED** [review]
Mir ist gerade noch aufgefallen: Angenommen ein Layer A und ein Layer B haben beide eine eigene Nextcloud-Instanz, und ich bin Adressverwalter o.ä. im Layer B. Kann ich mich dann im Nextcloud von Layer A einloggen und werde dort als Adressverwalter (entsprechend mit sehr hohen Berechtigungen)? Oder haben wir irgendeinen Mechanismus um das einzuschränken? Bisher gab es eine solche Möglichkeit eben nicht, und jetzt wo wir die Leute animieren, auch höhere Rollen im Nextcloud zu vergeben, wäre das schon noch wichtig. Siehst du das anders?

**Comment by kronn** [comment]
> Mir ist gerade noch aufgefallen: Angenommen ein Layer A und ein Layer B haben beide eine eigene Nextcloud-Instanz, und ich bin Adressverwalter o.ä. im Layer B. Kann ich mich dann im Nextcloud von Layer A einloggen und werde dort als Adressverwalter (entsprechend mit sehr hohen Berechtigungen)? Oder haben wir irgendeinen Mechanismus um das einzuschränken? Bisher gab es eine solche Möglichkeit eben nicht, und jetzt wo wir die Leute animieren, auch höhere Rollen im Nextcloud zu vergeben, wäre das schon noch wichtig. Siehst du das anders?

Es gibt die Möglichkeit zur Einschränkung nicht. Das OAuth-Token weiss zwar zu welcher Person es gehört, aber die zugehörige OAuth-Applikation ist global und nicht an einen Layer gebunden. Man könnte jetzt versuchen, über die Nextcloud-URL und die Redirect-URLs einen Match herzustellen, aber das erscheint mir fehlerträchtig.

Um diese Einschränkung zu ermöglichen, müssten wir OAuth-Applikationen an einen Layer oder eine Gruppe binden. Dies würde ich aber gerne in einem eigenen Schritt/Issue/Ticket machen. Dann kann man auch die Anforderungen besser diskutieren und hat schonmal die Grundlage einer Integration.

Grundsätzlich werden nur Gruppenzugehörigkeiten vergeben. Wenn mehrere gleiche Rollen in Nextcloud unterschiedlich dargestellt werden sollen, müsste man aktuell ein Proc oder so definieren, der klar verschieden Gruppen erstellt.
